### PR TITLE
Don't generate unnecessary rmeta files.

### PR DESCRIPTION
**Pull request by nnethercote**
As per https://github.com/rust-lang/rust/pull/60006#issuecomment-484284191.

r? @alexcrichton

### CODE REVIEW

**Comment by nnethercote** [comment]
@alexcrichton: remove the first rmeta generation as you suggested worked fine. The second one (conditional on it not being a dylib) didn't work... maybe I got the condition wrong? Details are in the comment in the code.

File: src/librustc_codegen_ssa/base.rs [hunk 1]
```diff
@@ -639,7 +638,18 @@ pub fn codegen_crate<B: ExtraBackendMethods>(
...
+    // njn: if I uncomment this condition I get tons of "can't find crate" test
+    // failures, like this:
+    //
+    // error[E0463]: can't find crate for `three_equals`
+    //   --> $DIR/three-equals.rs:6:1
+    //    |
+    //    | extern crate three_equals;
+    //    | ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate
+    //
+    //if tcx.sess.crate_types.borrow().contains(&config::CrateType::Dylib) {
```
Comment (alexcrichton) [inline src/librustc_codegen_ssa/base.rs]: Oh you know I think I lied when I said that we only put compressed metadata in `Dylib` crate types. I think that we actually also put it in `ProcMacro` crate types, and that may fix the test error you're seeing?

**Comment by nnethercote** [comment]
@alexcrichton New code is up!

**Comment by nnethercote** [comment]
@bors r-

@alexcrichton: I realize now that this change should do more -- the creation of `metadata_module` can be skipped in a lot of cases. I will fix that.

**Comment by nnethercote** [comment]
I ended up only making a tiny change -- I moved the creation of `metadata_module` later, so it's only created after the early `return`.

I had thought a bigger change was possible, that I could do the same thing with `metadata`, but it turns out the first `start_async_codegen` call needs `metadata`.

@bors r=alexcrichton

**Comment by oli-obk** [comment]
This PR broke clippy's test suite. We are using compiletest and are some of our ui tests use `serde`. Since this PR we're getting 

> error: extern location for serde does not exist: target/debug/deps/libserde-61a62cbc4921f441.rmeta

I'm assuming this is a compiletest bug or misconfiguration on our side. Any tips? Do we even need that file (like is it just wrongly generating the need for it)? Or do we need to pass a flag so it is generated again?

**Comment by ehuss** [comment]
@oli-obk Can you check if https://github.com/rust-lang/rust-clippy/pull/4115 fixes your issue?

**Pull request merged**